File: evals/refs/prho.zig
const std = @import("std");

fn gcdU(a_in: u64, b_in: u64) u64 {
    var a = a_in; var b = b_in;
    while (b != 0) {
        const t = a % b; a = b; b = t;
    }
    return a;
}

fn mulmod(a: u64, b: u64, m: u64) u64 {
    return @intCast((@as(u128, a) * @as(u128, b)) % @as(u128, m));
}

fn pollard(n: u64) u64 {
    if (n % 2 == 0) return 2;
    var c: u64 = 1;
    while (true) {
        var x: u64 = 2; var y: u64 = 2; var d: u64 = 1;
        while (d == 1) {
            x = (mulmod(x, x, n) + c) % n;
            y = (mulmod(y, y, n) + c) % n;
            y = (mulmod(y, y, n) + c) % n;
            const diff = if (x > y) x - y else y - x;
            d = gcdU(diff, n);
        }
        if (d != n) return d;
        c += 1;
    }
}

pub fn main(init: std.process.Init) !void {
    const arena = init.arena.allocator();
    var in_buf: [4096]u8 = undefined;
    var fr = std.Io.File.stdin().reader(init.io, &in_buf);
    const r = &fr.interface;
    var data: std.ArrayList(u8) = .empty;
    try r.appendRemainingUnlimited(arena, &data);
    var out_buf: [4096]u8 = undefined;
    var fw = std.Io.File.stdout().writer(init.io, &out_buf);
    const w = &fw.interface;
    var it = std.mem.tokenizeAny(u8, data.items, " \n\r\t");
    const t = try std.fmt.parseInt(usize, it.next().?, 10);
    var k: usize = 0;
    while (k < t) : (k += 1) {
        const n = try std.fmt.parseInt(u64, it.next().?, 10);
        try w.print("{d}\n", .{pollard(n)});
    }
    try w.flush();
}

File: evals/refs/prho.rs
use std::io::{self, Read, Write, BufWriter};

fn gcdu(mut a: u64, mut b: u64) -> u64 {
    while b != 0 { let t = a % b; a = b; b = t; } a
}
fn mulmod(a: u64, b: u64, m: u64) -> u64 {
    ((a as u128 * b as u128) % m as u128) as u64
}
fn pollard(n: u64) -> u64 {
    if n % 2 == 0 { return 2; }
    let mut c: u64 = 1;
    loop {
        let (mut x, mut y, mut d) = (2u64, 2u64, 1u64);
        while d == 1 {
            x = (mulmod(x, x, n) + c) % n;
            y = (mulmod(y, y, n) + c) % n;
            y = (mulmod(y, y, n) + c) % n;
            let diff = if x > y { x - y } else { y - x };
            d = gcdu(diff, n);
        }
        if d != n { return d; }
        c += 1;
    }
}

fn main() {
    let mut s = String::new();
    io::stdin().read_to_string(&mut s).unwrap();
    let stdout = io::stdout();
    let mut out = BufWriter::new(stdout.lock());
    let mut it = s.split_ascii_whitespace();
    let t: usize = it.next().unwrap().parse().unwrap();
    for _ in 0..t {
        let n: u64 = it.next().unwrap().parse().unwrap();
        writeln!(out, "{}", pollard(n)).unwrap();
    }
}
